ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,507 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
 static LIST_ENTRY PspQuotaBlockList = {&PspQuotaBlockList, &PspQuotaBlockList};
 static KSPIN_LOCK PspQuotaLock;
 
-#define TAG_QUOTA_BLOCK 'bQsP'
 #define VALID_QUOTA_FLAGS (QUOTA_LIMITS_HARDWS_MIN_ENABLE | \
                            QUOTA_LIMITS_HARDWS_MIN_DISABLE | \
                            QUOTA_LIMITS_HARDWS_MAX_ENABLE | \
                            QUOTA_LIMITS_HARDWS_MAX_DISABLE)
 
 /* PRIVATE FUNCTIONS *******************************************************/
 
-/*
- * Private helper to charge the specified process quota.
- * Returns STATUS_QUOTA_EXCEEDED on quota limit check failure.
- * Updates QuotaPeak as needed for specified quota type in PS_QUOTA_TYPE enum.
- * Notes: Conceptually translation unit local/private.
+/**
+ * @brief
+ * Returns pool quotas back to the Memory Manager
+ * when the pool quota block is no longer being
+ * used by anybody.
+ *
+ * @param[in] QuotaBlock
+ * The pool quota block which quota resources
+ * are to be sent back.
+ *
+ * @return
+ * Nothing.
+ */
+VOID
+NTAPI
+PspReturnQuotasOnDestroy(
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock)
+{
+    ULONG PsQuotaTypeIndex;
+    SIZE_T QuotaToReturn;
+
+    /* Make sure that the quota block is not plain garbage */
+    ASSERT(QuotaBlock);
+
+    /* Loop over the Process quota types */
+    for (PsQuotaTypeIndex = PsQuotaTypes; PsQuotaTypeIndex <  PsQuotaTypes; PsQuotaTypeIndex++)
+    {
+        /* The amount needed to return to Mm is the limit and return fields */
+        QuotaToReturn = QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Limit + QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Return;
+        MmReturnPoolQuota(PsQuotaTypeIndex, QuotaToReturn);
+
+        /*
+         * If that's a page file quota then don't
+         * do anything here, so just skip it outright.
+         */
+        if (PsQuotaTypeIndex == PsPageFile)
+        {
+            break;
+        }
+    }
+}
+
+/**
+ * @brief
+ * Releases some of excess quotas in order to
+ * attempt to free up some resources. This is
+ * done primarily in case the Memory Manager
+ * fails to raise the quota limit.
+ *
+ * @param[in] QuotaType
+ * Process pool quota type.
+ *
+ * @param[out] ReturnedQuotas
+ * A pointer to the returned amount of quotas
+ * back to Memory Manager.
+ *
+ * @return
+ * Nothing.
+ *
+ * @remarks
+ * The function releases excess paged or non
+ * paged pool quotas. Page file quota type is
+ * not permitted.
+ */
+VOID
+NTAPI
+PspReturnExcessQuotas(
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _Outptr_ PSIZE_T ReturnedQuotas)
+{
+    PPSP_QUOTA_LIST PspQuotaList;
```

Comment:
Ah, so that's the place where `PPSP_QUOTA_LIST` type is used. My suggestion: either move that typedef into that quota.c file, or just use here a `PLIST_ENTRY` straightforwardly (the latter seems to be the best, since all the usage done with it below clearly shows you're using such LIST_ENTRY).